File: examples/hooks/post-tool-trim.mjs
#!/usr/bin/env node
/**
 * PostToolUse — hard-trim oversized tool results (simple demo)
 *
 * Builtin compress_tool_result (headroom) is smarter; this shows how a
 * command hook can still rewrite modified_result itself.
 *
 * {
 *   "hooks": {
 *     "events": {
 *       "PostToolUse": [{
 *         "matcher": "Bash|View|Grep",
 *         "hooks": [{
 *           "type": "command",
 *           "command": "node examples/hooks/post-tool-trim.mjs",
 *           "timeout_ms": 3000,
 *           "fail_mode": "open"
 *         }]
 *       }]
 *     }
 *   }
 * }
 */

import { readEvent, allow, modifyResult } from "./lib/read-event.mjs";

const MAX_CHARS = Number(process.env.SOLCODE_HOOK_TRIM_CHARS || 12_000);

const event = await readEvent();
const block = event.tool_result;

if (!block || typeof block !== "object") {
  allow();
  process.exit(0);
}

// Never touch images / errors
if (block.type === "image" || block.is_error) {
  allow();
  process.exit(0);
}

const text = typeof block.text === "string" ? block.text : "";
if (text.length <= MAX_CHARS) {
  allow();
  process.exit(0);
}

const head = Math.floor(MAX_CHARS * 0.7);
const tail = MAX_CHARS - head - 80;
const trimmed =
  text.slice(0, head) +
  `\n\n... [post-tool-trim: omitted ${text.length - MAX_CHARS} chars] ...\n\n` +
  text.slice(-Math.max(tail, 0));

modifyResult(
  {
    type: "text",
    text: trimmed,
    is_error: Boolean(block.is_error),
  },
  `trimmed tool result ${text.length}→${trimmed.length} chars`,
);

File: examples/hooks/python/post_tool_trim.py
#!/usr/bin/env python3
"""PostToolUse — hard-trim oversized tool results (Python demo)."""

from __future__ import annotations

import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from lib.hook_io import allow, modify_result, read_event  # noqa: E402

MAX_CHARS = int(os.environ.get("SOLCODE_HOOK_TRIM_CHARS", "12000"))


def main() -> None:
    event = read_event()
    block = event.get("tool_result")
    if not isinstance(block, dict):
        allow()
        return
    if block.get("type") == "image" or block.get("is_error"):
        allow()
        return

    text = block.get("text") if isinstance(block.get("text"), str) else ""
    if len(text) <= MAX_CHARS:
        allow()
        return

    head = MAX_CHARS * 7 // 10
    tail = max(MAX_CHARS - head - 80, 0)
    trimmed = (
        text[:head]
        + f"\n\n... [post_tool_trim: omitted {len(text) - MAX_CHARS} chars] ...\n\n"
        + (text[-tail:] if tail else "")
    )
    modify_result(
        {
            "type": "text",
            "text": trimmed,
            "is_error": bool(block.get("is_error")),
        },
        f"trimmed tool result {len(text)}→{len(trimmed)} chars",
    )


if __name__ == "__main__":
    main()
